Implement stopping of matches

Malaber opened the pull request:
[TURNIERE-190](https://jiratinf16b4.it.dh-karlsruhe.de:8443/browse/TURNIERE-190)

Malaber committed:
Generalize tournament save interactor

It now is responsible for saving all ApplicationRecord objects to the
database. This will reduce code duplication one we have other objects
that need to be saved. (As we will soon need to save individual matches)

Malaber committed:
Remove unused evaluate status method

Malaber committed:
Change status of Match to only represent being finished

The winner will be a attribute of the Match instead of something
that is written in its status.

Malaber committed:
Add winner method to match

Malaber committed:
Call populate_match_below when a match is finished

Malaber committed:
Test stopping of Matches

Malaber committed:
Implement populate_match_below method

Malaber committed:
Test performance optimization

In general this generates less unnecessary faker data

Malaber committed:
Implement PopulateMatchBelow Interactor & Organizer

Malaber committed:
Check for errors while populating the match below

Malaber committed:
Fix second match winner advancing, even if match is not finished

Malaber committed:
Improve Formatting in playoff_stage_service (I guess)

... Rubocop wanted it that way

Malaber committed:
Rearrange new methods into one to split it later on

Malaber committed:
Rearrange methods in playoff_stage_service

Malaber committed:
Move populating the match below to when the state is changed

Malaber committed:
Change the way, existing match_scores are handled

Malaber committed:
Save match_scores explicitly after changing them

Malaber committed:
Test populate_match_below method

Malaber committed:
Actually make singleton_methods private

(Rubocop told me so and I trust the police)

Malaber committed:
Make save_application_record_object work with arrays

Malaber committed:
Return changed objects in populate_match_below

Malaber committed:
Test populate_match_below Interactor

Malaber committed:
Save match_scores after editing them in tests

turnieregitlabci reviewed (changes requested):
Blocks merging, This pull request is just symbolic for the changes it applies. It should actually point to master, but the factory changes are not on there yet, so this would not work.

Thor77 commented on app/models/match.rb:
```diff
@@ -19,23 +19,25 @@ def owner
-    end
+  private
+
+  def evaluate_winner
```
why is this a separate method and not in `.winner`?

Malaber replied:
Because I wanted to implement caching or writing it into database at first, then didn't manage to do so and decided to move that to another point in time.

Thor77 replied:
with caching (which is probably the better approach for this) the implementation will not change... therefore there's no need for two separate methods

Thor77 commented on app/services/playoff_stage_service.rb:
```diff
@@ -70,4 +70,73 @@ def self.calculate_required_stage_count(number_of_teams)
+    [match_below, match_scores].flatten
+  end
+
+  class << self
```
if you're defining this block here, please put all your static methods inside of it and not only private ones **or** rewrite the service into a module [ref](https://github.com/rubocop-hq/ruby-style-guide#modules-vs-classes)

Malaber replied:
done

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -5,6 +5,10 @@
     @match = create(:match, state: :not_started)
+    @match_owner = @match.owner
+    @tournament = create(:group_stage_tournament, stage_count: 3)
+    @running_playoff_match = @tournament.stages.find { |s| s.level == 3 }.matches.first
```
```suggestion
    @running_playoff_match = @tournament.stages.find_by(level: 3).matches.first
```

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -5,6 +5,10 @@
 RSpec.describe MatchesController, type: :controller do
   before do
     @match = create(:match, state: :not_started)
+    @match_owner = @match.owner
```
why? this is not even shorter than `@match.owner`...

Malaber replied:
Because then it gets cached and we don't have to look it up every time

Thor77 replied:
Rails is already doing that:
```
irb(main):003:0> m = Match.first
  Match Load (0.3ms)  SELECT  "matches".* FROM "matches" ORDER BY "matches"."id" ASC LIMIT ?  [["LIMIT", 1]]
=> #<Match id: 1, state: "single_team", position: nil, stage_id: nil, group_id: 1, created_at: "2019-05-08 07:19:42", updated_at: "2019-05-08 07:19:42">
irb(main):004:0> m.owner
  Group Load (0.4ms)  SELECT  "groups".* FROM "groups" WHERE "groups"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  Stage Load (0.4ms)  SELECT  "stages".* FROM "stages" WHERE "stages"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  Tournament Load (0.5ms)  SELECT  "tournaments".* FROM "tournaments" WHERE "tournaments"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  User Load (0.5ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
=> #<User id: 1, provider: "email", uid: "xochitl@barrowslang.com", allow_password_change: false, username: "sharda", email: "xochitl@barrowslang.com", created_at: "2019-05-08 07:19:42", updated_at: "2019-05-08 07:19:42">
irb(main):005:0> m.owner
=> #<User id: 1, provider: "email", uid: "xochitl@barrowslang.com", allow_password_change: false, username: "sharda", email: "xochitl@barrowslang.com", created_at: "2019-05-08 07:19:42", updated_at: "2019-05-08 07:19:42">
irb(main):006:0> m.owner
=> #<User id: 1, provider: "email", uid: "xochitl@barrowslang.com", allow_password_change: false, username: "sharda", email: "xochitl@barrowslang.com", created_at: "2019-05-08 07:19:42", updated_at: "2019-05-08 07:19:42">
```

Malaber replied:
done

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -5,6 +5,10 @@
+    @match_owner = @match.owner
+    @tournament = create(:group_stage_tournament, stage_count: 3)
+    @running_playoff_match = @tournament.stages.find { |s| s.level == 3 }.matches.first
+    @running_playoff_match_owner = @running_playoff_match.owner
```
why?

Malaber replied:
fixed

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -65,6 +69,40 @@
       end
     end
 
+    context 'as another owner' do
```
why are you creating another top-level context (with a confusing name) here?
imo this should be a sub-context of `context 'with valid params'` named `'on a running playoff match'`

Malaber replied:
it was easier, gonna look at it again

Malaber replied:
done

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -65,6 +69,40 @@
+        apply_authentication_headers_for @running_playoff_match_owner
+      end
+
+      context 'stops the match' do
```
no.

Malaber replied:
fixed (i guess)

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -65,6 +69,40 @@
+          expect(@running_playoff_match.state).to eq(finished[:state])
+        end
+
+        it 'fills the match below' do
```
```suggestion
        it 'populates the match below' do
```

Thor77 commented on spec/factories/tournaments.rb:
```diff
@@ -17,7 +17,7 @@
         group_count { 2 }
```
reminder: will hopefully go away once you rebase this branch on top of #37

Malaber replied:
(which it already is based on)

Thor77 commented on spec/interactors/populate_match_below_interactor_spec.rb:
```diff
@@ -0,0 +1,43 @@
+      expect(context).to be_a_success
+    end
+
+    it 'provides the obects to save' do
```
```suggestion
    it 'provides the objects to save' do
```

Thor77 commented on spec/models/match_spec.rb:
```diff
@@ -43,6 +43,23 @@
+
+    it 'returns a winner Team for a decided match' do
+      winner = @decided_match.winner
+      expect(winner).to be_a Team
```
you should probably check if it's returning the correct team instead of any...

Malaber replied:
Done

Thor77 commented on spec/models/match_spec.rb:
```diff
@@ -43,6 +43,23 @@
     end
   end
 
+  context '#winner' do
```
additionally there are 4 unnecessary var-assignments in this context /o\

Malaber replied:
I can't see a single unecessary assignment. No variable is only used only once or not changed after assigning

Thor77 replied:
https://github.com/turniere/turniere-backend/pull/38#pullrequestreview-240485750

Malaber replied:
forget it I am blind

Malaber replied:
done

Thor77 commented on spec/services/playoff_stage_service_spec.rb:
```diff
@@ -80,4 +80,78 @@
     end
   end
+
+  describe '# populate_match_below' do
```
```suggestion
  describe '#populate_match_below' do
```

Thor77 reviewed (changes requested):
just some style comments for now..

Malaber committed:
Simplify code finding the tournament

Co-Authored-By: Thor77 <thor77@thor77.org>

Thor77 commented on spec/models/match_spec.rb:
```diff
@@ -43,6 +43,23 @@
 
+  context '#winner' do
+    before do
+      @undecided_match = create(:undecided_group_match)
```
1

Malaber replied:
Well okay I apparently can't read, i looked at the other context.. sorry; gonna change it

Malaber replied:
done

Thor77 commented on spec/models/match_spec.rb:
```diff
@@ -43,6 +43,23 @@
+  context '#winner' do
+    before do
+      @undecided_match = create(:undecided_group_match)
+      @decided_match = create(:decided_playoff_match)
```
2

Malaber replied:
done

Thor77 commented on spec/models/match_spec.rb:
```diff
@@ -43,6 +43,23 @@
+    end
+
+    it 'returns a winner Team for a decided match' do
+      winner = @decided_match.winner
```
3

Malaber replied:
done

Thor77 commented on spec/models/match_spec.rb:
```diff
@@ -43,6 +43,23 @@
+    end
+
+    it 'returns nil for an undecided match' do
+      winner = @undecided_match.winner
```
4

Malaber replied:
done

Malaber committed:
Fix factory name for stage tournament

Malaber committed:
Fix factory name for stageless tournament

Malaber committed:
Move stopping of matches test

Malaber committed:
Move all methods in playoff_stage_service into self block

Malaber committed:
Rearrange Test Code

Malaber committed:
Change match state of match below

Thor77 reviewed (dismissed):
I will approve this for now, but we definitely have to revisit our service/interactor model.. this is far too much code and overhead for such a simple feature.

Malaber committed:
Check if .winner returns a Team

This is done to prevent Test from succeeding if both
.winner and .teams return nil

Malaber committed:
Force save match_scores

Malaber committed:
Move sonar-scanner to bottom of travis scripts

This is necessary to correctly submit coverage result to sonarqube

Thor77 reviewed (dismissed):
I will approve this for now, but we definitely have to revisit our service/interactor model.. this is far too much code and overhead for such a simple feature.

Malaber committed:
Create :stage_tournaments with realistic match positions

Malaber committed:
Fix comparison that got lost in refactoring

Thor77 reviewed (approved):
I will approve this for now, but we definitely have to revisit our service/interactor model.. this is far too much code and overhead for such a simple feature.
